internal/querycoordv2/checkers/balance_checker.go

```diff
@@ -18,34 +18,128 @@ package checkers
 
 import (
 	"context"
+	"sort"
 
+	"github.com/milvus-io/milvus/internal/log"
+	"github.com/milvus-io/milvus/internal/proto/querypb"
 	"github.com/milvus-io/milvus/internal/querycoordv2/balance"
+	"github.com/milvus-io/milvus/internal/querycoordv2/meta"
 	. "github.com/milvus-io/milvus/internal/querycoordv2/params"
+	"github.com/milvus-io/milvus/internal/querycoordv2/session"
 	"github.com/milvus-io/milvus/internal/querycoordv2/task"
+	"github.com/milvus-io/milvus/internal/util/typeutil"
+
+	"github.com/samber/lo"
+	"go.uber.org/zap"
 )
 
 // BalanceChecker checks the cluster distribution and generates balance tasks.
 type BalanceChecker struct {
 	baseChecker
 	balance.Balance
+	meta                            *meta.Meta
+	nodeManager                     *session.NodeManager
+	balancedCollectionsCurrentRound typeutil.UniqueSet
+	scheduler                       task.Scheduler
 }
 
-func NewBalanceChecker(balancer balance.Balance) *BalanceChecker {
+func NewBalanceChecker(meta *meta.Meta, balancer balance.Balance, nodeMgr *session.NodeManager, scheduler task.Scheduler) *BalanceChecker {
 	return &BalanceChecker{
-		Balance: balancer,
+		Balance:                         balancer,
+		meta:                            meta,
+		nodeManager:                     nodeMgr,
+		balancedCollectionsCurrentRound: typeutil.NewUniqueSet(),
+		scheduler:                       scheduler,
 	}
 }
 
 func (b *BalanceChecker) Description() string {
 	return "BalanceChecker checks the cluster distribution and generates balance tasks"
 }
 
-func (b *BalanceChecker) Check(ctx context.Context) []task.Task {
-	ret := make([]task.Task, 0)
+func (b *BalanceChecker) collectionsToBalance() []int64 {
+	ids := b.meta.GetAll()
+
+	// loading collection should skip balance
+	loadedCollections := lo.Filter(ids, func(cid int64, _ int) bool {
+		collection := b.meta.GetCollection(cid)
+		return collection != nil && b.meta.GetCollection(cid).Status == querypb.LoadStatus_Loaded
+	})
+	sort.Slice(loadedCollections, func(i, j int) bool {
+		return loadedCollections[i] < loadedCollections[j]
+	})
+
+	// balance collections influenced by stopping nodes
+	stoppingCollections := make([]int64, 0)
+	for _, cid := range loadedCollections {
+		replicas := b.meta.ReplicaManager.GetByCollection(cid)
+	outer:
+		for _, replica := range replicas {
+			for _, nodeID := range replica.GetNodes() {
+				isStopping, _ := b.nodeManager.IsStoppingNode(nodeID)
+				if isStopping {
+					stoppingCollections = append(stoppingCollections, cid)
+					break outer
```

Comment:
maybe we shall not break here. By default all loaded collections will be affected by a stopping node if no resource group is used. If we break here, all collection will need multiple round to recover.